agol-publish/NightStocker.py

```diff
@@ -0,0 +1,520 @@
+        tags: comma-separated string of tags (string)
+        description: AGOL description (string)
+        terms_of_use: AGOL terms of use/license info (string)
+        credits: AGOL Credits/Attribution (string)
+        folder: AGOL org's folder to move item to
+    protect: if True, set AGOL flag to prevent item from being deleted
+
+    returns: the published feature layer's itemid
+    '''
+
+    print("uploading")
+    sd_item = gis.content.add({}, data=service_definition)
+
+    #: Publishing
+    print("publishing")
+    published_item = sd_item.publish()
+
+    #: Updating information
+    print("sharing") #: Everyone and groups.
+    published_item.share(everyone=True, org=True, groups=info['groups'])
+    if protect:
+        print("delete protection")
+        published_item.protect(enable=True)
+    # sd_item.protect(enable=True)
+    # print('authoritative')
+    # published_item.content_status = 'authoritative'
+
+    print("updating info")
+    published_item.update(item_properties={
+        'tags': info['tags'],
+        'description': info['description'],
+        'licenseInfo': info['terms_of_use'],
+        'snippet': info['summary'],
+        'accessInformation': info['credits']
+        })
+
+    print('folder')
+    published_item.move(info['folder'])
+    sd_item.move(info['folder'])
+
+    #: Allow Downloads
+    print("downloads")
+    manager = arcgis.features.FeatureLayerCollection.fromitem(published_item).manager
+    manager.update_definition({ 'capabilities': 'Query,Extract' })
+
+    return published_item.itemid
+
+
+def create_service_definition(layer_info, sde_path, temp_dir, project_path, 
+                              map_name, describe):
+    '''Create a service defintion for a layer to be uploaded to AGOL from an
+    SDE using an existing ArcGIS Pro project.
+    
+    Parameters:
+    layer_info: Dictionary of info about the layer to be prepped for upload
+        fc_name: Fully qualified name of the SDE feature class to be uploaded 
+                 (string)
+        title: title of the item for AGOL (string)
+    sde_path: Path to the source .sde connection file
+    temp_dir: Directory for holding reprojected fgdb and .sddraft & .sd files
+    project_path: Path to an existing ArcGIS Pro project
+    map_name: Name of the map in the Pro project to use
+    describe: results of arcpy.da.Describe() on feature class
+
+    returns: path to the .sd file
+    '''
+
+    try:
+        start = datetime.datetime.now()
+
+        sgid_table = os.path.join(sde_path, layer_info['fc_name'])
+        is_table = describe['datasetType'] == 'Table'
+
+        projected_table = project_data(sgid_table, temp_dir, 'tempfgdb.gdb',
+                                       is_table)
+
+        #: Get project and map
+        proj = arcpy.mp.ArcGISProject(project_path)
+        for m in proj.listMaps():
+            if m.name == map_name:
+                agol_map = m
+        del m
```

Comment:
I want access to the `proj` object within this function so that I can `.save()` it. I intentionally get a new reference to the project each time to ensure that I've got a clean workspace. I think the current PR speak for this is 'an overabundance of caution' :)